rpc/types/src/macros.rs

```diff
@@ -17,35 +15,35 @@
 /// See the [`crate::json`] module for example usage.
 ///
 /// # Macro internals
-/// This macro has 2 branches with almost the same output:
-/// 1. An empty `Request` type
-/// 2. An `Request` type with fields
+/// This macro uses:
+/// - [`__define_request`]
+/// - [`__define_response`]
+/// - [`__define_request_and_response_doc`]
 ///
-/// The first branch is the same as the second with the exception
-/// that if the caller of this macro provides no fields, it will
-/// generate:
+/// # `__define_request`
+/// This macro has 2 branches. If the caller provides
+/// `Request {}`, i.e. no fields, it will generate:
 /// ```
 /// pub type Request = ();
 /// ```
-/// instead of:
+/// If they _did_ specify fields, it will generate:
 /// ```
 /// pub struct Request {/* fields */}
 /// ```
-///
 /// This is because having a bunch of types that are all empty structs
 /// means they are not compatible and it makes it cumbersome for end-users.
 /// Really, they semantically are empty types, so `()` is used.
 ///
-/// Again, other than this, the 2 branches do (should) not differ.
+/// # `__define_response`
+/// This macro has 2 branches. If the caller provides `Response`
+/// it will generate a normal struct with no additional fields.
+///
+/// If the caller provides a base type from [`crate::base`], it will
+/// flatten that into the request type automatically.
 ///
-/// FIXME: there's probably a less painful way to branch here on input
-/// without having to duplicate 80% of the macro. Sub-macros were attempted
-/// but they ended up unreadable. So for now, make sure to fix the other
-/// branch as well when making changes. The only de-duplicated part is
-/// the doc generation with [`define_request_and_response_doc`].
+/// E.g. `Response {/*...*/}` and `ResponseBase {/*...*/}`
+/// would trigger the different branches.
 macro_rules! define_request_and_response {
```

Comment:
The type generator macro no longer has duplicated code, it only has 1 branch and uses sub-macros instead.

Since we got rid of the RPC access `Request` stuff, this macro _always_ takes `Request {}` as input now. No fields will create the `type Request = ()`, where fields will create a normal request.

Responses still use the RPC access syntax, so it's one of:
- `Request` - adds nothing (some requests only have `status`, so we need this)
- `ResponseBase` - adds `base: crate::base::ResponseBase`
- `AccessResponseBase` - adds `base: crate::base::AccessResponseBase`